### PR TITLE
Feature configurable grant types

### PR DESCRIPTION
**Pull request by angelakis**
As discussed in https://github.com/IdentityPython/oidcendpoint/pull/59#issuecomment-693260033, I split up the configurable grant types feature (originally by-grant-types) from the token exchange specific changes to make review easier.

This branch refactors token_coop module to allow configuring different handlers for each grant type configured.

### CODE REVIEW

**Comment by rohe** [comment]
EndpointHelper is not a general class for all endpoints so I think we should name it TokenEndpointHelper .

File: src/oidcendpoint/oidc/token_coop.py
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py]: Remove this function

File: src/oidcendpoint/oidc/token_coop.py
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py]: Move this to separate function (`validate_state`)?

File: src/oidcendpoint/oidc/token_coop.py
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py]: Add `True` as a default value?

File: src/oidcendpoint/oidc/token_coop.py
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py:255]: Improve error message?

Check if the required fields have been initialized and then initiate the class

File: tests/test_35_oidc_token_coop_endpoint.py
Comment (nsklikas) [inline tests/test_35_oidc_token_coop_endpoint.py:210]: Check exception message

File: tests/test_35_oidc_token_coop_endpoint.py
Comment (nsklikas) [inline tests/test_35_oidc_token_coop_endpoint.py:186]: Add test for a custom grant?

File: src/oidcendpoint/oidc/token_coop.py
Comment (angelakis) [reply src/oidcendpoint/oidc/token_coop.py]: ok

**Comment by angelakis** [comment]
> EndpointHelper is not a general class for all endpoints so I think we should name it TokenEndpointHelper .

Correct, I changed it.

File: src/oidcendpoint/oidc/token_coop.py
Comment (angelakis) [reply src/oidcendpoint/oidc/token_coop.py:255]: OK I broke this part in multiple checks and exception handers.

File: src/oidcendpoint/oidc/token_coop.py
Comment (angelakis) [reply src/oidcendpoint/oidc/token_coop.py]: Added True as a default value and removed string "default". To be more consistent, I also added False, which just ignores that particular grant type and doesn't enable it.

As a result, we can now use None or True to use the default grant type class. This may seem a bit counter-intuitive, but IMO makes sense, as our configs are YAML and I believe the following configuration options should be allowed for a better UX and should mean the same thing:
```
grant_types_supported:
  - authorization_code

grant_types_supported:
  - authorization_code: True
```

File: src/oidcendpoint/oidc/token_coop.py
Comment (angelakis) [reply src/oidcendpoint/oidc/token_coop.py]: ok

**Comment by angelakis** [comment]
Should I also change token_coop module and class name to Token and replace the existing one? Or leave it for the future?

**Comment by rohe** [comment]
I think you should.

**Comment by angelakis** [comment]
> I think you should.

Previously the token endpoint was named AccessToken instead of Token, which is the correct name now. I guess we should try to keep AccessToken (maybe as an alias) for backwards compatibility?

**Comment by rohe** [comment]
I don't know. It is the token endpoint not the access token endpoint.
But sure keep AccessToken as an alias.

**Review by peppelinux — COMMENTED** [review]
LGTM
@angelakis is it time to merge it?

**Review by nsklikas — APPROVED** [review]

**Comment by angelakis** [comment]
> LGTM
> @angelakis is it time to merge it?

I wanted to rename to Token and create an alias for AccessToken but haven't found the time yet.

We can merge once I make the changes @nsklikas pointed today, the renaming can be done on a separate pr.

File: tests/test_35_oidc_token_coop_endpoint.py
Comment (angelakis) [reply tests/test_35_oidc_token_coop_endpoint.py:210]: Did that (and also improved the exception handling in regards to that).

File: tests/test_35_oidc_token_coop_endpoint.py
Comment (angelakis) [reply tests/test_35_oidc_token_coop_endpoint.py:186]: Added a custom grant test.

**Comment by angelakis** [comment]
I believe this can be merged now.

**Comment by peppelinux** [comment]
@rohe do we have to merge the new storage model before merge this or we can procede as It Is?

**Comment by peppelinux** [comment]
This looks ready for merge. My fingers are on the merge Button, Is there any doubts? Can we proceed?

**Comment by rohe** [comment]
Go ahead!

**Pull request merged**
